File: src/uniquePaths/unique.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # store results of subproblems
        dp = [[0 for _ in range(n)] for _ in range(m)]

        for i in range(m):
            for j in range(n):
                if 0 in [i, j]:
                    dp[i][j] = 1
                else:
                    # any given grid index can be reached either from the top or left
                    dp[i][j] = dp[i][j - 1] + dp[i - 1][j]

        return dp[m-1][n-1]


if __name__ == '__main__':
    m, n = 7, 3
    print(f"Input: m = {m}, n = {n}")
    print(f"Output: {Solution().uniquePaths(m, n)}")

File: src/uniquePaths/Unique.java
package uniquePaths;

class Unique {
    public int uniquePaths(int rows, int columns) {
        if (rows == 1 && columns == 1) {
            return 1;
        }
        int[][] dp = new int[rows][columns];
        helper(rows, columns, 0, 0, dp);
        return dp[0][0];
    }

    public boolean isValidSquare(int rows, int columns, int i, int j) {
        return i >= 0 && j >= 0 && i < rows && j < columns;
    }

    public int helper(int rows, int columns, int i, int j, int[][] dp) {
        if (!isValidSquare(rows, columns, i + 1, j) && !isValidSquare(rows, columns, i, j + 1)) {
            return 1;
        }
        if (dp[i][j] != 0) {
            return dp[i][j];
        }

        int sum = 0;
        if (isValidSquare(rows, columns, i + 1, j)) {
            sum += helper(rows, columns, i + 1, j, dp);
        }
        if (isValidSquare(rows, columns, i, j + 1)) {
            sum += helper(rows, columns, i, j + 1, dp);
        }
        dp[i][j] = sum;

        return sum;
    }

    public static void main(String[] args) {
        int m = 7, n = 3;
        System.out.println("Input: m = " + m + " n = " + n);
        System.out.println("Output: " + new Unique().uniquePaths(m, n));
    }
}